fhir-tools/src/main/java/com/ibm/fhir/tools/CodeGenerator.java

```diff
@@ -1475,6 +1477,154 @@ public int compare(JsonObject first, JsonObject second) {
             }
             cb.annotation("Constraint", valueMap);
         }
+                
+        // Generate constraint annotations from extensible and preferred bindings
+        generateVocabularyConstraints(structureDefinition, cb, className, lastId);
+    }
+    
+    /**
+     * Generates constraint annotations from extensible and preferred bindings.
+     * 
+     * @param structureDefinition
+     *            the structure definition
+     * @param cb
+     *            the code builder
+     * @param className
+     *            the class name
+     * @param lastId
+     *            the ID of the last contraint
+     */
+    private void generateVocabularyConstraints(JsonObject structureDefinition, CodeBuilder cb, String className, String lastId) {
+
+        // Collect elements that have an extensible or preferred binding
+        for (JsonObject elementDefinition : getElementDefinitions(structureDefinition).stream().filter(e -> !isProhibited(e)).filter(e -> hasExtensibleBinding(e)
+                || hasPreferredBinding(e)).collect(Collectors.toList())) {
+
+            String path = elementDefinition.getString("path");
+            String elementName = getElementNameWithoutPrefix(elementDefinition, path, className);
+            JsonObject binding = getBinding(elementDefinition);
+            String location = elementName.contains(".") ? elementName.replace(".div", ".`div`").replace("[x]", "") : "(base)";
+            String expressionElementName = "(base)".equals(location) ? elementName : "$this";
+
+            // Generate constraint for value set binding
+            String valueSet = binding.getString("valueSet", null);
+            if (valueSet != null) {
+                lastId = generateNextConstraintId(lastId, className);
+
+                String level = "Warning";
+                String description = (hasExtensibleBinding(elementDefinition) ? "SHALL, if possible, at least contain a code from value set "
```

Comment:
wondering if you need the 'at least' part in the description - I think 'SHALL, if possible, contain a code from value set' and 'SHOULD contain a code from value set' works

Reply:
I can remove the words "at least".